src/SqlBinding/TriggerBinding/SqlTableWatcher.cs

```diff
@@ -0,0 +1,604 @@
+                        while (await reader.ReadAsync())
+                        {
+                            _rows.Add(SqlBindingUtilities.BuildDictionaryFromSqlRow(reader, cols));
+                        }
+                    }
+                }
+
+                foreach (var row in _rows)
+                {
+                    // Not great that we're doing a SqlCommand per row, should batch this
+                    using (SqlCommand acquireLeaseCommand = BuildAcquireLeaseOnRowCommand(row, connection, transaction))
+                    {
+                        await acquireLeaseCommand.ExecuteNonQueryAsync();
+                        // Necessary so that other commands can re-use the parameters in the _primaryKeyValuesOfRows map
+                        acquireLeaseCommand.Parameters.Clear();
+                    }
+                }
+                await transaction.CommitAsync();
+            }
+        }
+
+        /// <summary>
+        /// Renews the leases held on _rows
+        /// </summary>
+        /// <returns></returns>
+        private async Task RenewLeasesAsync()
+        {
+            using (SqlConnection connection = SqlBindingUtilities.BuildConnection(_connectionStringSetting, _configuration))
+            {
+                await connection.OpenAsync();
+                foreach (var row in _rows)
+                {
+                    // Not great that we're doing a SqlCommand per row, should batch this
+                    using (SqlCommand renewLeaseCommand = BuildRenewLeaseOnRowCommand(row, connection))
+                    {
+                        await renewLeaseCommand.ExecuteNonQueryAsync();
+                        renewLeaseCommand.Parameters.Clear();
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Releases the leases held on _rows
+        /// </summary>
+        /// <returns></returns>
+        private async Task ReleaseLeasesAsync()
+        {
+            // Don't want to change the _rows while another thread is attempting to renew leases on them
+            await _renewLeasesLock.WaitAsync();
+            try
+            {
+                using (SqlConnection connection = SqlBindingUtilities.BuildConnection(_connectionStringSetting, _configuration))
+                {
+                    await connection.OpenAsync();
+                    SqlTransaction transaction = connection.BeginTransaction(System.Data.IsolationLevel.RepeatableRead);
+                    foreach (var row in _rows)
+                    {
+                        // Not great that we're doing a SqlCommand per row, should batch this
+                        using (SqlCommand releaseLeaseCommand = BuildReleaseLeaseOnRowCommand(row, connection, transaction))
+                        {
+                            await releaseLeaseCommand.ExecuteNonQueryAsync();
+                            releaseLeaseCommand.Parameters.Clear();
```

Comment:
Why clear the parameters collection if this object is going to be garbage collected anyways?